### PR TITLE
feat/frame: add HandshakeAckFrame

### PR DESCRIPTION
**Pull request by woorui**

### CODE REVIEW

**Review by fanweixiao — COMMENTED** [review]
for `core/server.go`, shall we response `HandshakeACK` on Line 166: https://github.com/yomorun/yomo/pull/401/files#diff-e2355be4a556783aaa497d1c8df81bcea308bef0b8fb1fc48b76ec561a891963R166

`s.handleConnection()` can be started after `HandshakeACK`.

@woorui @wujunzhuo

File: core/server.go [hunk 1]
```diff
@@ -163,6 +164,10 @@ func (s *Server) Serve(ctx context.Context, conn net.PacketConn) error {
 				}
 				defer stream.Close()
 
+				if ok := s.handshakeWithtimeout(conn, stream, 10*time.Second); !ok {
+					return
```
Comment (fanweixiao) [inline core/server.go:168]: Suggest close connection with an error msg of handshake timeout

File: core/server.go [hunk 2]
```diff
@@ -174,6 +179,56 @@ func (s *Server) Serve(ctx context.Context, conn net.PacketConn) error {
...
+
+	select {
+	case <-time.After(timeout):
+		return false
+	case ok := <-ch:
+		return ok
+	}
+}
+
+// handshake wait a handshakeFrame from client until timeout reached.
```
Comment (fanweixiao) [inline core/server.go:200]: should delete `until timeout reached.` here

File: core/server.go (hunk 1 above)
Comment (fanweixiao) [reply core/server.go:168]: or, a `GoawayFrame`?

File: core/server.go [hunk 3]
```diff
@@ -174,6 +179,56 @@ func (s *Server) Serve(ctx context.Context, conn net.PacketConn) error {
...
+	frm, err := fs.ReadFrame()
+	if err != nil {
+		if err := fs.WriteFrame(frame.NewGoawayFrame(err.Error())); err != nil {
+			logger.Errorf("%s⛔️ write to client[%s] GoawayFrame error:%v", ServerLogPrefix, conn.RemoteAddr().String(), err)
+		}
+		return false
+	}
+
+	if frm.Type() != frame.TagOfHandshakeFrame {
+		if err := fs.WriteFrame(frame.NewRejectedFrame("handshake failed")); err != nil {
```
Comment (fanweixiao) [inline core/server.go:214]: Is it possible return a `GoawayFrame` instead of `RejectedFrame`? minimize the variety of types of frame.

**Review by fanweixiao — APPROVED** [review]

File: core/server.go (hunk 1 above)
Comment (woorui) [reply core/server.go:168]: The code that response `goaway` or `ack` is placed in `handshake` function scope.

**Review by fanweixiao — APPROVED** [review]
👍🏼

File: core/frame/frame.go [hunk 4]
```diff
@@ -3,8 +3,71 @@ package frame
...
+	// and frame returned is nil.
+	ReadFrame() (Frame, error)
+}
+
+// ErrReadUntilTimeout be returned when call ReadUntil timeout.
+type ErrReadUntilTimeout struct{ t Type }
+
+// Error implement error interface.
+func (err ErrReadUntilTimeout) Error() string {
+	return "yomo: frame read until timeout, tpye: " + err.t.String()
```
Comment (venjiang) [inline core/frame/frame.go:27]: tpye -> type

File: core/frame/frame.go [hunk 5]
```diff
@@ -3,8 +3,71 @@ package frame
 import (
 	"os"
 	"strconv"
+	"time"
 )
 
+// Readwriter is the interface that groups the ReadFrame and WriteFrame methods.
+type Readwriter interface {
```
Comment (venjiang) [inline core/frame/frame.go:10]: Readwriter should be `ReadWriter`

**Review by wujunzhuo — APPROVED** [review]

File: core/server.go [hunk 6]
```diff
@@ -163,6 +164,10 @@ func (s *Server) Serve(ctx context.Context, conn net.PacketConn) error {
 				}
 				defer stream.Close()
 
+				if ok := s.handshakeWithtimeout(conn, stream, 10*time.Second); !ok {
```
Comment (venjiang) [inline core/server.go:167]: should be `handshakeWithTimeout`

**Review by venjiang — APPROVED** [review]

**Pull request merged**